File: 1700-minimum-time-to-make-rope-colorful/minimum-time-to-make-rope-colorful.kt
class Solution {
    fun minCost(colors: String, neededTime: IntArray): Int {
        val n: Int = colors.length;

        var cost: Int = 0;
        for (i in 1 until n) {
            if (colors[i] == colors[i - 1]) {
                cost += min(neededTime[i], neededTime[i - 1]);
                neededTime[i] = max(neededTime[i], neededTime[i - 1]);
            }
        }

        return cost;
    }
}

File: 1700-minimum-time-to-make-rope-colorful/minimum-time-to-make-rope-colorful.cpp
class Solution {
public:
    int minCost(string colors, vector<int>& neededTime) {
        neededTime.push_back(10001);
        colors.push_back('@');
        int n = neededTime.size();

        int cost = 0;
        for (int i = 1, sum = neededTime[0]; i < n; i++) {
            if (colors[i] == colors[i - 1]) {
                sum += neededTime[i];
                neededTime[i] = max(neededTime[i], neededTime[i - 1]);
            } else {
                cost += sum - neededTime[i - 1];
                sum = neededTime[i];
            }
        }

        return cost;
    }
};